subprojects/base-services/src/test/groovy/org/gradle/internal/reflect/JavaReflectionUtilTest.groovy

```diff
@@ -62,12 +62,36 @@ class JavaReflectionUtilTest extends Specification {
 
     def "write property"() {
         when:
-        writeableProperty(JavaTestSubject, "myProperty").setValue(myProperties, "otherValue")
+        writeableProperty(JavaTestSubject, "myProperty", String.class).setValue(myProperties, "otherValue")
 
         then:
         readableProperty(JavaTestSubject, String, "myProperty").getValue(myProperties) == "otherValue"
     }
 
+    def "picks the generic object setter if the typed setter does not match the value type"() {
+        when:
+        def property = writeableProperty(JavaTestSubject, "myProperty", File.class)
+
+        then:
+        property.type == Object.class
+    }
+
+    def "picks the generic iterable setter if the typed setter does not match the value type"() {
+        when:
+        def property = writeableProperty(JavaTestSubject, "multiValue", List.class)
+
+        then:
+        property.type == Iterable.class
+    }
+
+    def "can handle null as property type"() {
```

Comment:
❌  `null` behaviour feels underspecified here.

The test should make clear which setter gets called, specially when there's more than a possible null receiver (like with the `multiValue` property, for instance).

Another question that comes to mind: What's the effect of passing `null` to a property of a primitive type?